docs/static_site/src/pages/api/cpp/docs/tutorials/multi_threaded_inference.md

```diff
@@ -0,0 +1,324 @@
+4. Prepare lambda function which will run in spawned threads. Call C API to **invoke cached op** within the lambda function.
+5. Spawn multiple threads and wait for all threads to complete.
+6. Post process data to obtain inference results and cleanup.
+
+### Step 1: Parse arguments and load input image into ndarray
+
+```c++
+int main(int argc, char *argv[]) {
+  if (argc < 5) {
+    std::cout << "Please provide a model name, num_threads, is_gpu, test_image" << std::endl
+              << "Usage: ./multi_threaded_inference [model_name] [num_threads] [is_gpu] apple.jpg"
+              << std::endl
+              << "Example: ./.multi_threaded_inference imagenet1k-inception-bn 1 0 apple.jpg"
+              << std::endl
+              << "NOTE: Thread number ordering will be based on the ordering of file inputs" << std::endl
+              << "NOTE: Epoch is assumed to be 0" << std::endl;
+    return EXIT_FAILURE;
+  }
+  std::string model_name = std::string(argv[1]);
+  int num_threads = std::atoi(argv[2]);
+  bool is_gpu = std::atoi(argv[3]);
+  ...
+  ...
+  mxnet::cpp::Shape input_shape = mxnet::cpp::Shape(1, 3, 224, 224);
+  for (size_t i = 0; i < files.size(); i++) {
+    files[i].resize(image_size);
+    GetImageFile(test_files[i], files[i].data(), channels,
+                 cv::Size(width, height));
+    input_arrs.emplace_back(mxnet::cpp::NDArray(files[i].data(), input_shape, mxnet::cpp::Context::cpu(0)));
+  }
+```
+
+The above code parses arguments, loads the image file into a ndarray with a specific shape. There are a few things that are set by default and not configurable. For example, `static_alloc` and `static_shape` are by default set to true.
+
+
+### Step 2: Prepare input data and load parameters, copying data to a specific context
+```c++
+void run_inference(const std::string& model_name, const std::vector<mxnet::cpp::NDArray>& input_arrs,
+                   std::vector<mxnet::NDArray*> *output_mx_arr,
+                   int num_inf_per_thread = 1, bool random_sleep = false,
+                   int num_threads = 1, bool static_alloc = false,
+                   bool static_shape = false,
+                   bool is_gpu = false) {                                                                                       
+  ...
+  ...
+  ...
+  // Prepare input data and parameters
+  std::vector<mxnet::cpp::NDArray> data_arr(num_threads);
+  std::vector<mxnet::cpp::NDArray> softmax_arr;
+  std::vector<mxnet::cpp::NDArray> params;
+  mxnet::cpp::Shape data_shape = mxnet::cpp::Shape(1, 3, 224, 224);
+  mxnet::cpp::Shape softmax_shape = mxnet::cpp::Shape(1);
+  int num_inputs = out.ListInputs().size();
+
+  for (size_t i = 0; i < data_arr.size(); ++i) {
+    data_arr[i] = input_arrs[i].Copy(ctx);
+  }
+  prepare_input_data(softmax_shape, ctx, num_threads, &softmax_arr);
+  std::map<std::string, mxnet::cpp::NDArray> parameters;
+  mxnet::cpp::NDArray::Load(param_file, 0, &parameters);
+
+  for (std::string name : out.ListInputs()) {
```

Comment:
reference?  does this code compile? Can be included in the doc? how do we keep it current.

Reply:
no, this is a doc and I had added this to refer to the code that I was refering to in the cc file. I changed it to the code blocks in the cc file, to make it easier to maintain.